### ISSUE TITLE
Cache for preOutput()

### ISSUE THREAD

**Issue by raver119** [deeplearning4j/deeplearning4j#3385]
In CNN architecture significant amount of time is spent within preOutput() method, which gets called twice: once in FF pass,  and later in BP pass.

We could cache output of this method in host memory during FF pass. We already have concept of "external"  workspace, that holds activations/gradients during iteration, we could introduce concept similar to that: stash. I.e. k/v storage limited to HOST-only memory, so once we store some INDArray there - it gets copied to host memory. Once we need to get that INDArray back - we copy it back from storage.

**Comment by raver119** [comment]
@AlexDBlack

**Comment by AlexDBlack** [comment]
Following up on our gitter conversation:
The whole "do forward pass twice" design was decided some time ago to minimise memory requirements.

In some cases like dense layers, the additional memory requirement is equal to the activation size. However in the case of CNNs and sub sampling layers, also need the im2col arrays, which can be much much larger than the activation arrays by factor of four (for 2x2 conv layers stride 1) to nine (3x3 stride 1) or even more some cases.

If we implement this, I think we should also give users the option to prioritise memory or speed.

Anyway I'll go through this and quantify the exact memory requirements, and exactly what arrays we need to keep in memory each layer type.

**Comment by AlexDBlack** [comment]
DenseLayer and OutputLayer:
- Performance: we save doing ```input.mmul(W).addiRowVector(b);``` that's ```[minibatch,nIn]x[nIn,nOut]``` for the mmul
- Memory: increased by ```[minibatch,nOut]```

Convolution layer:
- Memory:
    - Need to save im2col array: ```[miniBatch, outH, outW, inDepth, kH, kW]```. For same mode, stride 1, (worst case) this equates to ```(kH*kW)*inputSize```. In general for same mode, I think it becomes ```(kH*kW)*inputSize/(sH*sW)``` so for the common setting (k=3,s=1) and (k=3,s=2) we have 9x and 2.25x input size respectively.
    - Need to store pre-out: increases memory by ```[minibatch, outDepth, outH, outW]```
- Performance:
   - We save a costly im2col op
   - We save an mmul op: ```[minibatch*outH*outW, (inDepth*kH*kW)] x [(inDepth*kH*kW), outDepth]``` and an addiRowVector

Subsampling layer:
- looks quite similar to convolution layer case. We save an im2col and a max/avg op, with the same im2col memory cost.


My main concern at this point are the size of those im2col arrays.

**Comment by raver119** [comment]
But please note: cuDNN helper doesn't provide im2col, so it seems there's only 1 array to be stored?

**Comment by raver119** [comment]
Stash was stupid idea, and won't be implemented. It will be done without any additional entities introduced, just workspace that's allocated in HOST-memory only, since nd4j-cuda allows that. All we need is layer-wise variable, pushed to external workspace, that can be using either DEVICE or HOST memory. So simple as: if field is null on backpass - we call for preOutput. If field is not null - we consume it, and set to null. Just almost no additional magic is needed for it, and behavior will be configurable by user.

**Comment by crockpotveggies** [comment]
I'll take over implementation of this and get a prototype going. Should be easy as per @raver119 's suggestions

**Comment by raver119** [comment]
So, tests show that benefit is neglectable for small networks (i.e. LeNet, AlexNet) and quite significant on larger networks (i.e. ~20% for VGG-16).

So, implementation incoming.

**Comment by raver119** [comment]
So, some news and facts here...
With GPU RAM being used for cache stuff works fine, however introduces additional memory use.

With HOST-only memory being used for cache, on my box i've got significant slowdown. After short investigation, it turned out that PCIe (in my case 3.0 x8) is bottleneck. With HOST->GPU or GPU->HOST transfers i get approximately 6GB/s bandwidth, so, say, 150m elements array copy takes approx 98ms.

With PCIe x16 it will be twice faster, but still slower then required to keep up with desired performance.
So, basically desired transfer speeds are possible to achieve only with NVLink (80GB/s and above) or PCIe 4.0 (96GB/s)

**Comment by raver119** [comment]
After thinking about this issue, i'm adding configurable caching for preOutput:

1) NONE - preOutput won't be cached
2) DEVICE - preOutput will be cached in device memory (if applicable). 
3) HOST - preOutput will be cached in host memory.

So, basically for CUDA backend if user have enough memory for DEVICE cache - he can use it. If he's on NVLink hardware, he can use HOST cache, since it's not that slow as my PCIe x8.

And for CPU backend both options will reside in the same memory space, just different workspaces.

**Comment by raver119** [comment]
P.S. default is NONE.

**Comment by raver119** [comment]
All configurations and sample implementation were merged to master. Other layers to be adopted, will file new issue.
